src/core/loss_detection.c

```diff
@@ -1496,6 +1498,66 @@ QuicLossDetectionProcessAckBlocks(
     }
 
     if (NewLargestAck) {
+        if (Path->EcnValidationState != ECN_VALIDATION_FAILED) {
+            //
+            // Per RFC 9000, we validate ECN counts from received ACK frames
+            // when the largest acked packet number increases.
+            //
+            QUIC_PACKET_SPACE* Packets = Connection->Packets[EncryptLevel];
+            BOOLEAN EcnValidated = TRUE;
+            int64_t EctCeDeltaSum = 0;
+            if (Ecn != NULL) {
+                EctCeDeltaSum += Ecn->CE_Count - Packets->EcnCeCounter;
+                EctCeDeltaSum += Ecn->ECT_0_Count - Packets->EcnEctCounter;
+                //
+                // Conditions where ECN validation fails:
+                // 1. Reneging ECN counts from the peer.
+                // 2. ECN counts do not match the marks that were applied to the packets sent.
+                //
+                if (EctCeDeltaSum < 0 ||
+                    EctCeDeltaSum < EcnEctCounter ||
+                    Ecn->ECT_1_Count != 0 ||
+                    Connection->Send.NumPacketsSentWithEct < Ecn->ECT_0_Count) {
+                    EcnValidated = FALSE;
+                } else {
+                    //
+                    // TODO: Notify CC of the ECN signal before we update the CE counts.
+                    //
+                    Packets->EcnCeCounter = Ecn->CE_Count;
+                    Packets->EcnEctCounter = Ecn->ECT_0_Count;
+                    if (Path->EcnValidationState <= ECN_VALIDATION_UNKNOWN) {
+                        Path->EcnValidationState = ECN_VALIDATION_CAPABLE;
+                        QuicTraceLogConnInfo(
```

Comment:
Shouldn't this be the same `ConnEcnCapable` event as in the rundown?

Reply:
fixed